asf_tools/composite.py

```diff
@@ -0,0 +1,308 @@
+    if path:
+        logging.info("Searching for list of files to process")
+
+        # New format directory names
+        infiles_new = glob.glob(os.path.join(path, f"20*/S1?_IW_*RTC*/*{pol}.tif"))
+
+        # Old format diretory names
+        infiles_old = glob.glob(os.path.join(path, f"20*/PRODUCT/*{pol}.tif"))
+
+        infiles = infiles_new
+        infiles.extend(infiles_old)
+        cnt = len(infiles)
+        logging.info(f"Found {cnt} files to process")
+    else:
+        cnt = len(infiles)
+        logging.info("Found list of {cnt} input files to process")
+    infiles.sort()
+    logging.debug(f"Input files: {infiles}")
+
+    # resample infiles to maximum resolution & common UTM zone
+    resampled_files = reproject_to_median_utm(infiles, pol, resolution=resolution)
+    if len(resampled_files) == 0:
+        Exception("Unable to resample files")
+
+    # Get pixel size
+    x, y, trans, proj = saa.read_gdal_file_geo(saa.open_gdal_file(resampled_files[0]))
+    pixel_size_x = trans[1]
+    pixel_size_y = trans[5] 
+    logging.info(f"{resampled_files[0]} x = {pixel_size_x} y = {pixel_size_y}")
+
+    # Get extent of union of all images
+    extents = []
+    for fi in resampled_files:
+        ulx, lrx, lry, uly = saa.getCorners(fi)
+        extents.append([fi, ulx, lrx, lry, uly])
+    ulx, lrx, uly, lry = get_full_extent(extents)
+
+    logging.info(f"Full extent of mosaic is {ulx,uly} to {lrx,lry}")
+   
+    x_pixels = abs(int((ulx - lrx) / pixel_size_x))
+    y_pixels = abs(int((lry - uly) / pixel_size_y))
+
+    logging.info(f"Output size is {x_pixels} samples by {y_pixels} lines")
+
+    outputs = np.zeros((y_pixels, x_pixels))
+    weights = np.zeros((y_pixels, x_pixels))
+    counts = np.zeros((y_pixels, x_pixels), dtype=np.int8)
+    logging.info("Calculating output values")
+
+    for fi, x_max, x_min, y_max, y_min in extents:
+        if pol in fi:
+            logging.info(f"Processing file {fi}")
+            logging.info(f"File covers {x_max,y_min} to {x_min,y_max}")
+
+            logging.info("Reading areas")
+            x_size, y_size, trans, proj, areas = saa.read_gdal_file(saa.open_gdal_file(fi.replace(f"_{pol}_reproj",
+                                                                                                  "_area_reproj")))
+
+            logging.info("Reading values")
+            x_size, y_size, trans, proj, values = saa.read_gdal_file(saa.open_gdal_file(fi))
+
+            out_loc_x = (x_max - ulx) / pixel_size_x
+            out_loc_y = (y_min - uly) / pixel_size_y
+            end_loc_x = out_loc_x + x_size
+            end_loc_y = out_loc_y + y_size
+
+            logging.info(f"Placing values in output grid at {int(out_loc_x)}:{int(end_loc_x)} "
+                         f"and {int(out_loc_y)}:{int(end_loc_y)}")
+
+            temp = 1.0/areas 
+            temp[values == 0] = 0
+            mask = np.ones((y_size,x_size),dtype = np.uint8)
+            mask[values == 0] = 0
+
```

Comment:
using array operations and functions like [`numpy.clip`](https://numpy.org/doc/stable/reference/generated/numpy.clip.html) would significantly reduce the memory overhead here

Reply:
The code uses array operations and it doesn't seem to be clipping anything here.  It is creating a mask of all ones or zeros to increase the counts array for valid pixels only.  Where would one use clip in this code?

Reply:
Sorry, these were comments to myself mostly since it's still a draft PR. 

`np.clip` might be needed for `1.0/areas` as this will -> inf as area -> 0. 

These kind of statements
```
temp[values == 0] = 0
```
Are much more memory intensive than using something like, [`numpy.ma.masked_where`](https://numpy.org/doc/stable/reference/generated/numpy.ma.masked_where.html) (or other masking functions) because they need two copies of the array, and the masking functions can operate in place. 

Using these, and then later extending to a [Dask Array](https://docs.dask.org/en/latest/) would allow us to make global mosaics on a moderately sized machine.